pkg/network/encoding/http.go

```diff
@@ -54,116 +28,84 @@ func newHTTPEncoder(httpPayloads map[http.Key]*http.RequestStats) *httpEncoder {
 	}
 
 	return &httpEncoder{
+		builder: model.NewHTTPAggregationsBuilder(nil),
 		byConnection: GroupByConnection("http", httpPayloads, func(key http.Key) types.ConnectionKey {
 			return key.ConnectionKey
 		}),
-		aggregations: new(model.HTTPAggregations),
 	}
 }
 
-func (e *httpEncoder) GetHTTPAggregationsAndTags(c network.ConnectionStats) ([]byte, uint64, map[string]struct{}) {
+func (e *httpEncoder) GetHTTPAggregationsAndTags(c network.ConnectionStats, builder *model.ConnectionBuilder) (uint64, map[string]struct{}) {
 	if e == nil {
-		return nil, 0, nil
+		return 0, nil
 	}
 
 	connectionData := e.byConnection.Find(c)
 	if connectionData == nil || len(connectionData.Data) == 0 || connectionData.IsPIDCollision(c) {
-		return nil, 0, nil
+		return 0, nil
 	}
 
-	return e.encodeData(connectionData)
-}
+	var (
+		staticTags  uint64
+		dynamicTags map[string]struct{}
+	)
 
-func (e *httpEncoder) encodeData(connectionData *USMConnectionData[http.Key, *http.RequestStats]) ([]byte, uint64, map[string]struct{}) {
-	e.reset()
+	builder.SetHttpAggregations(func(b *bytes.Buffer) {
+		staticTags, dynamicTags = e.encodeData(connectionData, b)
+	})
+	return staticTags, dynamicTags
+}
 
+func (e *httpEncoder) encodeData(connectionData *USMConnectionData[http.Key, *http.RequestStats], w io.Writer) (uint64, map[string]struct{}) {
 	var staticTags uint64
 	dynamicTags := make(map[string]struct{})
+	e.builder.Reset(w)
 
 	for _, kvPair := range connectionData.Data {
-		key := kvPair.Key
-		stats := kvPair.Value
-
-		ms := httpStatsPool.Get().(*model.HTTPStats)
-		ms.Path = key.Path.Content
-		ms.FullPath = key.Path.FullPath
-		ms.Method = model.HTTPMethod(key.Method)
-		ms.StatsByStatusCode = e.getDataMap(stats.Data)
-
-		for status, s := range stats.Data {
-			data := ms.StatsByStatusCode[int32(status)]
-			data.Count = uint32(s.Count)
-
-			if latencies := s.Latencies; latencies != nil {
-				blob, _ := proto.Marshal(latencies.ToProto())
-				data.Latencies = blob
-			} else {
-				data.FirstLatencySample = s.FirstLatencySample
+		e.builder.AddEndpointAggregations(func(httpStatsBuilder *model.HTTPStatsBuilder) {
+			key := kvPair.Key
+			stats := kvPair.Value
+
+			httpStatsBuilder.SetPath(key.Path.Content)
+			httpStatsBuilder.SetFullPath(key.Path.FullPath)
+			httpStatsBuilder.SetMethod(uint64(model.HTTPMethod(key.Method)))
+
+			for code, stats := range stats.Data {
+				httpStatsBuilder.AddStatsByStatusCode(func(w *model.HTTPStats_StatsByStatusCodeEntryBuilder) {
+
+					if code != 0 {
+						w.SetKey(int32(code))
+					}
```

Comment:
why do we check `code != 0`? it is not possible as we're checking it earlier in the process